binary_search_tree/tree.py

```diff
@@ -3,57 +3,183 @@ def __init__(self, key, val = None):
         if val == None:
             val = key
 
-        self.key = key
         self.value = val
+        self.key = key
         self.left = None
         self.right = None
-        
-
-
+        # self.val = key
+    
 class Tree:
-    def __init__(self):
+    def __init__(self, height = None):
         self.root = None
+        self.nodes = []
+        #hold for the bfs
+        self.queque = []
+        # self.treeheight = 0
+        self.bfs_nodes = []
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n) because looping through all the nodes
+    # Space Complexity: O(n) because you need to add a node for each key, value
     def add(self, key, value = None):
-        pass
+        
+        if self.root == None:
+            self.root = TreeNode(key,value)
+        else:
+            parent = None
+            current = self.root
+            while (current !=None):
+                parent = current
+                if current.key > key:
+                    current = current.left
+                else:
+                    current = current.right
+            if parent.key > key:
+                parent.left = TreeNode(key,value)
+            else:
+                parent.right = TreeNode(key,value)
+        
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def create_dict(self, TreeNode):
+        return { "key": TreeNode.key,
+                 "value": TreeNode.value }
+        
+
+    # Time Complexity: log(n) because we are examining only half of the nodes
+    # Space Complexity: O(1) because we are not allocating any memory
+    # self refers to object the class is making
     def find(self, key):
```

Comment:
👍 The time complexity is O(n log n ) if the tree is balanced and O(n) otherwise.